contracts/Bridge.sol

```diff
@@ -322,51 +336,43 @@ contract Bridge is Pausable, AccessControl, SafeMath {
     function voteProposal(uint8 chainID, uint64 depositNonce, bytes32 resourceID, bytes32 dataHash) external onlyRelayers whenNotPaused {
 
         uint72 nonceAndID = (uint72(depositNonce) << 8) | uint72(chainID);
-        Proposal storage proposal = _proposals[nonceAndID][dataHash];
+        Proposal memory proposal = _proposals[nonceAndID][dataHash];
 
         require(_resourceIDToHandlerAddress[resourceID] != address(0), "no handler for resourceID");
         require(uint(proposal._status) <= 1, "proposal already passed/executed/cancelled");
-        require(!_hasVotedOnProposal[nonceAndID][dataHash][msg.sender], "relayer already voted");
-
-        if (uint(proposal._status) == 0) {
-            ++_totalProposals;
-            _proposals[nonceAndID][dataHash] = Proposal({
-                _resourceID : resourceID,
-                _dataHash : dataHash,
-                _yesVotes : new address[](1),
-                _noVotes : new address[](0),
-                _status : ProposalStatus.Active,
-                _proposedBlock : block.number
-                });
-
-            proposal._yesVotes[0] = msg.sender;
-            emit ProposalEvent(chainID, depositNonce, ProposalStatus.Active, resourceID, dataHash);
-        } else {
-            if (sub(block.number, proposal._proposedBlock) > _expiry) {
-                // if the number of blocks that has passed since this proposal was
-                // submitted exceeds the expiry threshold set, cancel the proposal
-                proposal._status = ProposalStatus.Cancelled;
-                emit ProposalEvent(chainID, depositNonce, ProposalStatus.Cancelled, resourceID, dataHash);
-            } else {
-                require(dataHash == proposal._dataHash, "datahash mismatch");
-                proposal._yesVotes.push(msg.sender);
-
-
-            }
+        require(!_hasVoted(proposal, msg.sender), "relayer already voted");
 
+        if (proposal._status == ProposalStatus.Inactive) {
+            proposal = Proposal({
+                _status : ProposalStatus.Active,
+                _yesVotes : 0,
+                _yesVotesTotal: 0,
+                _proposedBlock : uint40(block.number) // Overflow is desired.
```

Comment:
This is because we can rely on just the low order bits?

Reply:
Yeah, we have 40bits in storage for this. In case the block number on underlying blockchain will ever get over 40bits, say `(2**40-1) + 10`, then we will set **_proposedBlock** to 9. We only use it for expiration checks:
`sub(block.number, proposal._proposedBlock) > _expiry` the idea was that here, **block.number** will overflow too, and expression will still make sense. But now I see that it will not because after the overflow, it will always be bigger than expiry (unless we set expiry to a huge number).
Here is how am I planning to fix this:
`(uint40(block.number - proposal._proposedBlock) > _expiry`
Now we have 2 possible situations:
1. _proposedBlock not overflown yet -> normal subtraction because block.number >= _proposedBlock.
2. _proposedBlock overflown -> uint40 casting overflow (desired), because `block.number - _proposedBlock >= (2**40 - 1)`, and so after uint40 casting it will again tell us how many blocks have passed.

The problem that can still happen, is the case when some proposal gets so old (and still not cancelled) that `uint40(block.number - _proposedBlock)` overflows 2+ times. It means that proposal will not be considered as one that exceeded expiry threshold for some time again. Perhaps this is only a hypothetical possibility `(2**40 - 1)` blocks is 1,099,511,627,775. I don't think anyone can maliciously use it.